crates/voicevox_core_c_api/src/lib.rs

```diff
@@ -692,6 +716,253 @@ pub extern "C" fn voicevox_error_result_to_message(
     C_STRING_DROP_CHECKER.blacklist(message).as_ptr()
 }
 
+/// ユーザー辞書
+pub struct VoicevoxUserDict {
+    dict: Arc<Mutex<voicevox_core::UserDict>>,
+}
+
+/// ユーザー辞書の単語
+#[repr(C)]
+pub struct VoicevoxUserDictWord {
+    /// 表記
+    surface: *const c_char,
+    /// 読み
+    pronunciation: *const c_char,
+    /// アクセント型
+    accent_type: usize,
+    /// 単語の種類
+    word_type: VoicevoxUserDictWordType,
+    /// 優先度
+    priority: u32,
+}
+
+/// ユーザー辞書の単語の種類
+#[repr(i32)]
+#[allow(non_camel_case_types)]
+#[derive(Copy, Clone)]
+pub enum VoicevoxUserDictWordType {
+    /// 固有名詞。
+    VOICEVOX_USER_DICT_WORD_TYPE_PROPER_NOUN = 0,
+    /// 一般名詞。
+    VOICEVOX_USER_DICT_WORD_TYPE_COMMON_NOUN = 1,
+    /// 動詞。
+    VOICEVOX_USER_DICT_WORD_TYPE_VERB = 2,
+    /// 形容詞。
+    VOICEVOX_USER_DICT_WORD_TYPE_ADJECTIVE = 3,
+    /// 接尾辞。
+    VOICEVOX_USER_DICT_WORD_TYPE_SUFFIX = 4,
+}
+
+/// VoicevoxUserDictWordを最低限のパラメータで作成する。
+/// @param [in] surface 表記
+/// @param [in] pronunciation 読み
+/// @return VoicevoxUserDictWord
+///
+/// # Safety
+/// @param surface, pronunciation は有効な文字列へのポインタであること
+#[no_mangle]
+pub extern "C" fn voicevox_user_dict_word_make(
+    surface: *const c_char,
+    pronunciation: *const c_char,
+) -> VoicevoxUserDictWord {
+    VoicevoxUserDictWord {
+        surface,
+        pronunciation,
+        accent_type: UserDictWord::default().accent_type,
+        word_type: UserDictWord::default().word_type.into(),
+        priority: UserDictWord::default().priority,
+    }
+}
+
+/// ユーザー辞書を作成する
+/// @param [out] out_user_dict VoicevoxUserDictのポインタ
+/// @return 結果コード #VoicevoxResultCode
+///
+/// # Safety
+/// @param out_user_dict VoicevoxUserDictのポインタが有効な領域を指していること
+#[no_mangle]
+pub unsafe extern "C" fn voicevox_user_dict_new(
+    out_user_dict: NonNull<Box<VoicevoxUserDict>>,
+) -> VoicevoxResultCode {
+    let dict = voicevox_core::UserDict::new();
+    let user_dict = Box::new(VoicevoxUserDict {
+        dict: Arc::new(Mutex::new(dict)),
+    });
+    out_user_dict.as_ptr().write_unaligned(user_dict);
+
+    VoicevoxResultCode::VOICEVOX_RESULT_OK
+}
```

Comment:
エラーを返さないのならこうできると思います。

```suggestion
///
/// @return ユーザー辞書。
#[no_mangle]
pub extern "C" fn voicevox_user_dict_new() -> Box<VoicevoxUserDict> {
    // FIXME: ↑にある`struct VoicevoxUserDict`に`#[derive(Default)]`を付ける
    Default::default()
}
```